# HG changeset patch
# User Pantxo Diribarne <pantxo.diribarne@gmail.com>
# Date 1517844589 -3600
#      Mon Feb 05 16:29:49 2018 +0100
# Node ID 800876b7132c24dd6cd977f022a765f91515aad9
# Parent  dd68506c092603e401fd3fbb950d27f1688dd6d6
Make some markers linejoin/linecap round for better looking printout (bug #53023)

* gl2ps-print.cc (gl2ps_renderer::init_marker): reimplement method to change
  linecap and linejoin on a per marker basis.

diff --git a/libinterp/corefcn/gl2ps-print.cc b/libinterp/corefcn/gl2ps-print.cc
--- a/libinterp/corefcn/gl2ps-print.cc
+++ b/libinterp/corefcn/gl2ps-print.cc
@@ -172,16 +172,35 @@ namespace octave
     }
 
     void draw_text (const text::properties& props);
 
     void draw_pixels (int w, int h, const float *data);
     void draw_pixels (int w, int h, const uint8_t *data);
     void draw_pixels (int w, int h, const uint16_t *data);
 
+    void init_marker (const std::string& m, double size, float width)
+    {
+      opengl_renderer::init_marker (m, size, width);
+
+      // FIXME: gl2ps can't handle closed contours and we set linecap/linejoin
+      //        round to obtain a better looking result for some markers.
+      if (m == "o" || m == "v" || m == "^" || m == ">" || m == "<" || m == "h"
+          || m == "hexagram" || m == "p" || m == "pentagram")
+        {
+          set_linejoin ("round");
+          set_linecap ("round");
+        }
+      else
+        {
+          set_linejoin ("miter");
+          set_linecap ("square");
+        }
+    }
+
     void set_linestyle (const std::string& s, bool use_stipple = false,
                         double linewidth = 0.5)
     {
       opengl_renderer::set_linestyle (s, use_stipple, linewidth);
 
       if (s == "-" && ! use_stipple)
         gl2psDisable (GL2PS_LINE_STIPPLE);
       else
